### PR TITLE
feat(Translators): Support for DateAdd methods.

### LINKED ISSUE #6402: Support for DateAdd
### Steps to reproduce

I am trying to write a query where I want to return records where a child entity in a detail record has a data that is 10 years later than a field in it's parent. Instead of doing that, EF loops sending a bunch of additional queries.
### The issue

p is the main object, p.StatusChanges the child:

``` c#
qry = qry.Where(p => 
    p.ProjectStatusId == cid &&
    p.IsUnderLitigation && p.LitigationEndDate != null && 
    p.StatusChanges.Any(psc => psc.ToStatusId == cid && 
        psc.ModifiedOn <= p.LitigationEndDate.Value.AddYears(-10))));
```
### Further technical details

If I comment out the 

``` c#
    p.StatusChanges.Any(psc => psc.ToStatusId == cid && 
        psc.ModifiedOn <= p.LitigationEndDate.Value.AddYears(-10))
```

I get the correct query

``` SQL
SELECT [p].[Id], [p].[AjeraClientType], [p].[AlertText], [p].[BillingContactId], [p].[BillingSpecialInstructions], [p].[CPR], [p].[CTR], [p].[ClientId], [p].[ConsolidatedDate], [p].[Description], [p].[ECPR], [p].[ECTR], [p].[EFLocation], [p].[EmailInvoice], [p].[EstimatedFee], [p].[HasAlert], [p].[HasMarketingMaterials], [p].[HasNonBarcodedContractDocs], [p].[HasProjectSummary], [p].[HasRequiredNotes], [p].[IsDestroyedDateUnknown], [p].[IsPaperFileCreated], [p].[IsPaperFileMissing], [p].[IsPurgedDateUnknown], [p].[IsReoccurring], [p].[IsThinned], [p].[IsUnderLitigation], [p].[LastPaperFileMissingDate], [p].[Latitude], [p].[LitigationEndDate], [p].[LitigationStartDate], [p].[Longitude], [p].[MarketingFolderLocation], [p].[MarketingProjectNumber], [p].[ProjectBudget], [p].[ProjectHomeOfficeId], [p].[ProjectStatusId], [p].[PurchaseOrder], [p].[SpecialInstructions], [p].[SubmissionDate], [p].[Year], [p0].[Id], [p0].[ColorARGB], [p0].[Name], [c].[Id], [c].[Name], [c].[PreviousName]
FROM [Project] AS [p]
INNER JOIN [ProjectStatus] AS [p0] ON [p].[ProjectStatusId] = [p0].[Id]
INNER JOIN [Client] AS [c] ON [p].[ClientId] = [c].[Id]
WHERE (EXISTS (
    SELECT 1
    FROM [Folder] AS [f]
    WHERE ([f].[OfficeId] = @__q_OfficeId_Value_0) AND ([p].[Id] = [f].[ProjectId])) OR EXISTS (
    SELECT 1
    FROM [EFolder] AS [ef]
    WHERE ([ef].[OfficeId] = @__q_OfficeId_Value_1) AND ([p].[Id] = [ef].[ProjectId]))) AND ((([p].[ProjectStatusId] = 3000) AND (([p].[IsUnderLitigation] = 0) AND EXISTS (
    SELECT 1
    FROM [ProjectStatusChange] AS [psc]
    WHERE (([psc].[ToStatusId] = 3000) AND ([psc].[ModifiedOn] <= @__cutoff_2)) AND ([p].[Id] = [psc].[ProjectId])))) OR (([p].[IsUnderLitigation] = 1) AND [p].[LitigationEndDate] IS NOT NULL))
ORDER BY [p].[Id] DESC
OFFSET @__p_3 ROWS FETCH NEXT @__p_4 ROWS ONLY
```

If I uncomment it I get this 

``` SQL
SELECT [p].[Id], [p].[AjeraClientType], [p].[AlertText], [p].[BillingContactId], [p].[BillingSpecialInstructions], [p].[CPR], [p].[CTR], [p].[ClientId], [p].[ConsolidatedDate], [p].[Description], [p].[ECPR], [p].[ECTR], [p].[EFLocation], [p].[EmailInvoice], [p].[EstimatedFee], [p].[HasAlert], [p].[HasMarketingMaterials], [p].[HasNonBarcodedContractDocs], [p].[HasProjectSummary], [p].[HasRequiredNotes], [p].[IsDestroyedDateUnknown], [p].[IsPaperFileCreated], [p].[IsPaperFileMissing], [p].[IsPurgedDateUnknown], [p].[IsReoccurring], [p].[IsThinned], [p].[IsUnderLitigation], [p].[LastPaperFileMissingDate], [p].[Latitude], [p].[LitigationEndDate], [p].[LitigationStartDate], [p].[Longitude], [p].[MarketingFolderLocation], [p].[MarketingProjectNumber], [p].[ProjectBudget], [p].[ProjectHomeOfficeId], [p].[ProjectStatusId], [p].[PurchaseOrder], [p].[SpecialInstructions], [p].[SubmissionDate], [p].[Year]
FROM [Project] AS [p]
WHERE EXISTS (
    SELECT 1
    FROM [Folder] AS [f]
    WHERE ([f].[OfficeId] = @__q_OfficeId_Value_0) AND ([p].[Id] = [f].[ProjectId])) OR EXISTS (
    SELECT 1
    FROM [EFolder] AS [ef]
    WHERE ([ef].[OfficeId] = @__q_OfficeId_Value_1) AND ([p].[Id] = [ef].[ProjectId]))
```

Followed by a long series of these:

``` SQL
SELECT [psc1].[ProjectId]
FROM [ProjectStatusChange] AS [psc1]
WHERE ([psc1].[ToStatusId] = 3000) AND ([psc1].[ModifiedOn] <= @__cutoff_2)

SELECT [psc1].[ProjectId]
FROM [ProjectStatusChange] AS [psc1]
WHERE ([psc1].[ToStatusId] = 3000) AND ([psc1].[ModifiedOn] <= @__cutoff_2)

SELECT [psc1].[ProjectId]
FROM [ProjectStatusChange] AS [psc1]
WHERE ([psc1].[ToStatusId] = 3000) AND ([psc1].[ModifiedOn] <= @__cutoff_2)

SELECT [psc1].[ProjectId]
FROM [ProjectStatusChange] AS [psc1]
WHERE ([psc1].[ToStatusId] = 3000) AND ([psc1].[ModifiedOn] <= @__cutoff_2)
```

...

### PR DESCRIPTION
**Pull request by laskoviymishka**
**Please check if the PR fulfills these requirements**

- [x] The code builds and tests pass (verified by our automated build checks)
- [x] Commit messages follow this format
```
    Add translator for DataAdd SQL function
    - Translate DateTime.Add time function into sql DATEADD
    - Corresponding tests

    Fixes #6402
```
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code meets the expectations our engineering guidelines. 

Closes #6402

### CODE REVIEW

**Comment by anpete** [comment]
@laskoviymishka Thanks for the PR! Can you please rebase it on top of [feature/1.2.0](/aspnet/EntityFramework/tree/feature/1.2.0)? I fixed some inconsistencies with how we were dealing with SQL function expressions and so you will want to switch to using SqlFunctionExpression for this.

**Comment by laskoviymishka** [comment]
@anpete done. Please review.

**Comment by laskoviymishka** [comment]
@anpete done, please review.

File: src/Microsoft.EntityFrameworkCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerCompositeMethodCallTranslator.cs [hunk 1]
```diff
@@ -32,7 +32,8 @@ public class SqlServerCompositeMethodCallTranslator : RelationalCompositeMethodC
             new SqlServerStringToUpperTranslator(),
             new SqlServerStringTrimEndTranslator(),
             new SqlServerStringTrimStartTranslator(),
-            new SqlServerStringTrimTranslator()
+            new SqlServerStringTrimTranslator(),
+            new SqlServerDateAddTranslator()
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerCompositeMethodCallTranslator.cs]: Alphabetical sorting.

Comment (laskoviymishka) [inline src/Microsoft.EntityFrameworkCore.SqlServer/Query/ExpressionTranslators/Internal/SqlServerCompositeMethodCallTranslator.cs]: done

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs [hunk 2]
```diff
@@ -430,6 +431,7 @@ public override void Optional_navigation_inside_nested_method_call_translated_to
                 @"SELECT [e1].[Id], [e1].[Date], [e1].[Name], [e1].[OneToMany_Optional_Self_InverseId], [e1].[OneToMany_Required_Self_InverseId], [e1].[OneToOne_Optional_SelfId], [e1.OneToOne_Optional_FK].[Id], [e1.OneToOne_Optional_FK].[Date], [e1.OneToOne_Optional_FK].[Level1_Optional_Id], [e1.OneToOne_Optional_FK].[Level1_Required_Id], [e1.OneToOne_Optional_FK].[Name], [e1.OneToOne_Optional_FK].[OneToMany_Optional_InverseId], [e1.OneToOne_Optional_FK].[OneToMany_Optional_Self_InverseId], [e1.OneToOne_Optional_FK].[OneToMany_Required_InverseId], [e1.OneToOne_Optional_FK].[OneToMany_Required_Self_InverseId], [e1.OneToOne_Optional_FK].[OneToOne_Optional_PK_InverseId], [e1.OneToOne_Optional_FK].[OneToOne_Optional_SelfId]
 FROM [Level1] AS [e1]
 LEFT JOIN [Level2] AS [e1.OneToOne_Optional_FK] ON [e1].[Id] = [e1.OneToOne_Optional_FK].[Level1_Optional_Id]
+WHERE DATEADD(month, 2, DATEADD(day, 15E0, DATEADD(day, 10E0, [e1.OneToOne_Optional_FK].[Date]))) > '2002-02-01T00:00:00.000'
```
Comment (anpete) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs]: Looks to be because AddDays takes a double.

**Pull request merged**